### ISSUE TITLE
Implicit NotNull instrumentation should not apply to type parameters

### ISSUE THREAD

**Issue by yborovikov** [osundblad/intellij-annotations-instrumenter-maven-plugin#19]
Implicit not-null instrumentation should not be applied to not annotated parameters/return values whose type is a (not annotated) type parameter.

The reason is that the type arguments themselves can be annotated (with Java 8 type use annotations) by the caller himself to his preference (e.g. `List<@Nullable String>` vs. `List<@NotNull String>`) and intended nullability (from the caller point of view) may conflict with statically instrumented bytecode (and its implied not-null contract), causing false-positive `IllegalArgumentException`s and/or `NullPointerException`s being thrown.

E.g. this code:
```java
public class Holder<T> {
    private final T t;
    public Holder(T t) { this.t = t; }
    public T get() { return t; }
    public static void main(String[] args) {
        @Nullable Integer i = new Holder<@Nullable Integer>(null).get();
    }
}
```
will throw an `IllegalArgumentException` when `main` method is invoked (if built with `<implicit>true</implicit>` configuration of the plugin).

Unpack attached [intellij-annotations-instrumenter-maven-plugin-15.zip](https://github.com/osundblad/intellij-annotations-instrumenter-maven-plugin/files/1002518/intellij-annotations-instrumenter-maven-plugin-15.zip) and run `mvn clean process-classes && java -cp target/classes com.example.Holder` from its root directory to reproduce.

**Comment by osundblad** [comment]
After thinking a very short time: This is would be a really cool feature but I am not sure this is even possible with the current code base / annotations. The instrumenter modifies class files and there is only one class file here namely Holder so I think this would require that the annotation is kept to runtime and look up the annotation in each call which is not something I am planning (I want to do minimal changes to class files).

But please convince me that this is possible and I will look further into it. 

If this is a real issue for you there are some ways to solve it:
1 If you really need some nullable and some not: create two classes Holder and NullableHolder.
2 If not: annotate Holder constructor parameter and get return value as @Nullable  
3 Instead of annotating the methods as @Nullable it wrap the type with an Optional (which then is never null). Note make sure to store the null value internally and wrap it again when you return it since Optional is not Serializable (or use SOptional in my JType library). eg Optional<Integer> instead of @Nullable Integer.

**Comment by yborovikov** [comment]
Hi @osundblad,

I'm not asking for differentiation based on actual run-time annotations on the type parameters (that wouldn't be possible with Java generics erasure, as you mentioned).

What I am asking for - if a (callee) type parameter itself is not EXPLICITLY annotated (as either nullable or nonnull) - do not (implicitly) instrument it at all (as expected nullability is not defined in the instrumented class - it depends on caller-site annotations, if any).

Theoretically, such caller-annotated invocations could be instrumented at the caller site - but that would be a separate story :)

Looking at the code I realize that this change would require passing additional information about the argument/return type along the plugin call stack...

As for workarounds - currently I just don't use implicit instrumentation at all, that's not a big deal.

**Comment by atorstling** [comment]
Hi Yegor. If I understand you correctly, wouldn't your suggestion remove all implicit  instrumentations for fields of parameterized types? If so I think that will create a lot of confusion if you are not using call-site annotations, and also require those users to explicitly add annotations. 

Perhaps as an alternative we could introduce an annotation on those fields for skipping the implicit instrumentation?

**Comment by yborovikov** [comment]
Hi Alexander,

It would remove implicit instrumentation on parameters/return values whose type is an unannotated generic type parameter, e.g.:
```java
class Foo<@Nullable T1, @NotNull T2, T3> {

  T1 get1(); // T1 is known to be nullable - no instrumentation

  T2 get2(); // T2 is known to be non-null - instrument accordingly

  T3 get3(); // nullability of T3 is unknown at compile time - should not imply anything
             // since Java 8+ caller _can_ have its own type-use annotations on T3,
             // e.g. Foo<String, String, @Nullable String>, and implicit non-null
             // instrumentation on get3() would be inappropriate

}
```
What I am saying is it's ok (with `<implicit>true</implicit>` config) to assume unannotated parameter/return types to be not null when these types are known to be "stable", i.e. their nullability annotation (or lack thereof) is finalized at compile time.

This is not the case for type parameters - their actual type (and its nullability) is determined only once the type is fully parameterized.

You're right that such change would remove not-null instrumentation from generic class method parameters/return values which are typed with unannotated type parameters. I'd argue that that's the correct behavior.

Another example:
```java
interface Map<K, V> {
  ...
  boolean put(K key, V value); // generic map supports whatever K/V
  @Nullable V get(K key); // but it may return null if the key is not found
  ...
}

interface ConcurrentMap<K, V> extends Map<@NotNull K, @NotNull V> {
  // this interface takes only non-null keys and values
  // but get(K) still may return null
}
```

**Comment by yborovikov** [comment]
> Perhaps as an alternative we could introduce an annotation on those fields for skipping the implicit instrumentation?

Yes, that's an option. I'd rather use the lack of annotation as such hint, though.

Frankly, I prefer code to express its contract explicitly (to benefit both humans and static code analyzers) and was using implicit non-nulls more as a last resort for catching issues with the code that expects its arguments/return values to be non-null but does not express the contract.

Ideally, the plugin could use `@NonNullByDefault` annotations on packages for figuring out implicit nullability.
Of course, this is just my humble opinion, and I always have the choice of not requesting implicit instrumentation.

**Comment by osundblad** [comment]
Just to verify the idea, you'd like the following code to not throw an exception (note the place change of ```@Nullable```) when run?

```
@ImplictNotNull
public class Holder<@Nullable T> {
    private final T t;
    public Holder(T t) { this.t = t; }
    public T get() { return t; }
    public static void main(String[] args) {
        @Nullable Integer i = new Holder<Integer>(null).get();
    }
}
```
I think that might actually be possible to implement. 

To keep consistency I think that for:
```
class Foo<@Nullable T1, @NotNull T2, T3> 
``` 
should make T3 NotNull if implicit is true and Nullable otherwise. Also a method (return/parameter) annotation should override the Type annotation.


PS I wonder if I would use this since I am not sure that ```@Nullable``` would show if you go to a method (eg. Quick Doc/Def in IDEA) you'd have to check T at top of a class. But then again you would have fewer bugs when forgetting to annotate new methods using T.
PPS I actually wanted to put the ```@ImplicitNotNull``` in the ```package-info.java``` file, but since no class file is generated for it that became rather messy (since I had to have access to the source code and parse that).

**Comment by yborovikov** [comment]
Hi Olle,

Yes, `T` should be assumed to be nullable wherever it's used as a parameter/return type in that case (unless, again, explicitly overridden by local annotation). I didn't realize it's not the case currently :)
Static analysis tools (e.g. IntelliJ IDEA) are pretty good at making use of nullability contracts expressed that way.

As for consistency re unannotated type parameters, I'd argue that these are left unannotated explicitly to allow the consumer of the class to set its own nullability, hence assuming these to be non-null would be an issue (#19 : ). <sub>*That doesn't apply to normal types (not type parameters) - for those it's safe to apply implicit non-null contract since customer has no way to express its own expectations.</sub>

However, the simple workaround of not using implicit non-null configuration is available (though it cannot be applied to 3rd-party libs that were compiled with the implicit...).

Re `@ImplicitNotNull`: I believe a `package-info.class` is produced by `javac` and should be accessible at the plugin runtime?
